### ISSUE TITLE
`solc@0.8.30` not supported

### ISSUE THREAD

**Issue by bee344** [paritytech/revive#307]
`solc@0.8.30` was released and now is breaking compatibility with `resolc`, and therefore every tool that uses `@parity/resolc` is breaking with this error:
```bash
| ERROR: The command failed.
| Output:'Downloading compiler 0.8.28
| Compiling 1 Solidity file'
| Error: 'An unexpected error occurred:
| Error: `solc` versions >0.8.29 are not supported, found 0.8.30
```
Either we need to bind `solc` to >0.8.30 [here](https://github.com/paritytech/revive/blob/11d47d74ac2f4814abfe72fb3fad61a19f3042cf/js/resolc/package.json#L35) or update this [value](https://github.com/paritytech/revive/blob/11d47d74ac2f4814abfe72fb3fad61a19f3042cf/crates/solidity/src/solc/mod.rs#L22).

**Comment by smiasojed** [comment]
I think that the user utilizing the tooling should be able to specify the resolc version and the solc version that will be used for compilation. I believe that with the current resolc package design, we cannot achieve this. Do you consider refactoring this npm package?

**Comment by xermicus** [comment]
Thanks for reporting this. Yes support for resolc is explicit. We are not going to remove the last supported version. They can always break whatever they want and with every release we go over it and enable support for it manually (after studying release notes etc.).

**Comment by xermicus** [comment]
> Either we need to bind solc to >0.8.30

I guess you mean `<0.8.30`? I think the bug here is that the NPM package allows any semver compatible version while it should honor the versions support by resolc.

**Comment by xermicus** [comment]
> I think that the user utilizing the tooling should be able to specify the solc version that will be used for compilation. I believe that with the current resolc package design, we cannot achieve this. Do you consider refactoring this npm package?

Good point. But if the dependencies of this package specifies `<0.8.30` (for example). Anyone using it can just overwrite it with a more strict requirement and it would use that? I am not familiar with JS and dependencies but this is what I would expect. Same thing for just installing the binary: Can't the user just tell npm to install the `@parity/resolc` package but pin the `solc` package version to a specific one and then this just works without any changes?

**Comment by smiasojed** [comment]
You could probably override it, another option is to have an API that allows passing a soljson compiler object.

**Comment by bee344** [comment]
>I guess you mean <0.8.30? I think the bug here is that the NPM package allows any semver compatible version while it should honor the versions support by resolc.

Yeah, just a typo. The issue is the `^` that tells npm to use whatever version is compatible, and technically as long as it's a patch or a minor it is considered compatible. it's not a bug since that's how npm is designed to work, we should change it from our side.

>Can't the user just tell npm to install the @parity/resolc package but pin the solc package version to a specific one and then this just works without any changes?

Currently trying that with the `overrides` option of pnpm, but it doesn't seem like the package using `0.8.30` is in a reachable place from the tool side.

**Comment by xermicus** [comment]
> You could probably override it, another option is to have an API that allows passing a soljson compiler object.

Would mean that our package bundles every solc version we support? If so I don't think it is a sensible thing to do since it will install 30 versions?

> Currently trying that with the overrides option of pnpm, but it doesn't seem like the package using 0.8.30 is in a reachable place from the tool side.

ChatGPT is telling me about overrides too. Seems like what we want. I don't get what you mean with `it doesn't seem like the package using 0.8.30 is in a reachable place from the tool side`? We want the override to pin to 0.8.29.

**Comment by smiasojed** [comment]
> Would mean that our package bundles every solc version we support? If so I don't think it is a sensible thing to do since it will install 30 versions?

Soljson compilers are exposed over HTTP, so you can download any version you want and simply set it in resolc

**Comment by bee344** [comment]
>ChatGPT is telling me about overrides too. Seems like what we want. I don't get what you mean with it doesn't seem like the package using 0.8.30 is in a reachable place from the tool side? We want the override to restrict to 0.8.29. The override is just a quick fix to unbreak people and general solution if people want for some reason depend on some older version explicitly. Unless I am missing something here.

Never mind what I said, I wasn't applying the override to the tests' `package.json`. It works now.

**Comment by bee344** [comment]
If curious, you can see how the override was applied here https://github.com/paritytech/hardhat-polkadot/pull/107.

**Comment by xermicus** [comment]
@bee344 nice! This means we don't need to release something here immediately and at least hardhat users will be unblocked? The npm release is just part of the regular release job with everything. I want to do a new release here anyways in the coming days (but not just yet).

**Comment by xermicus** [comment]
However on a second thought I am not sure yet whether it is worth the trouble. Given overrides work fine, the only thing this breaks is users wanting to use the npm package as if it was a binary (Instead of the binary they install the npm package standalone without a package.json anywhere) and they want to use an older solc version than the latest support. Then they just have to have a package.json with an override somewhere I guess. Doesn't seem like a big deal at all.

**Comment by bee344** [comment]
Yeah, if overrides work it's fine I guess, it's not an issue that will be there permanently, since it depends on whatever the supported version is at the time, so it doesn't seem like going through the trouble of updating the version in the `package.json` every time is worth it.

**Comment by smiasojed** [comment]
The solc version is specified in hardhat config for the project. How is it enforced for resolc? Asking user to set some overrides does not look like user friendly solution.

**Comment by bee344** [comment]
Oh the end user doesn't need to worry about that, we make the overrides directly inside of the plugin.`@parity/resolc` doesn't provide the option to choose which `solc` version you want to use.

**Comment by xermicus** [comment]
Ah okay so hardhat folks did not implement this properly. Instead of using the package manager to handle dependencies the hand rolled it for solc. That's annoying because eventually our package here will need to replicate that.

**Comment by bee344** [comment]
Yes, we should have the overrides be defined by the user, I have to see how to remove that from the package itself and have the plugin use the user defined one.

**Comment by xermicus** [comment]
Closed by #308
